### PR TITLE
drivers/can: stm32_fd: move clock configuration from kconfig to devicetree

### LINKED ISSUE #49890: drivers/can: stm32_fd: CONFIG_CAN_STM32FD_CLOCK_DIVISOR not applied in driver setup
**Describe the bug**
The configuration of the `st,stm32_canfd` controller clock divider via  CONFIG_CAN_STM32FD_CLOCK_DIVISOR does not work.

**To Reproduce**
- Compile any can project for a target with a `st,stm32_canfd` compatible can controller and change CONFIG_CAN_STM32FD_CLOCK_DIVISOR to a value other than `1`.
- You will notice that the rate returned by `get_core_clock` changes accoridingly, while the configuration is not applied to the `FDCAN_CONFIG->CKDIV` register.

**Expected behavior**
Clock divider is applied correctly and `get_core_clock` returns correct frequency, such that I am able to configure the correct can bit rates.

**Impact**
- all devices using `st,stm32_canfd` like STM32G4 and STM32U5 SoCs
  - can core clock divider is not applied.
  - As get_core_clock returns the rate as if the divider would have been applied correctly the wrong can bit rate may be configured.

**Environment:**
 - Commit SHA: a5ee862ff0

### PR DESCRIPTION
**Pull request by str4t0m**
- Use domain clocks if they are defined in devicetree.

    Until now domain clock sources could be selected via kconfig.
    STM32 platform now can configure domain clock sources via
    clock control driver, therefore this commit makes use of it.
   
    The configuration is shared between canfd instances, so a domain clock
    has to be defined for only one instance. Otherwise, only the
    configuration from the latest initialized instance will remain.

- move clock divider configuration to devicetree

    Remove the CAN_STM32FD_CLOCK_DIVISOR configuration option,
    and add configuration via dts property clk-divider instead.

-   Fix clock divider configuration

    The configurable CAN clock divider CAN_STM32FD_CLOCK_DIVISOR
    was not applied during initialization, because write protection
    was not disabled.
    
    While the clock divider was not applied, it was still used in clock rate
    calculation, therefore resulting in incorrect bus speed setup.

Fixes #49890

### CODE REVIEW

**Comment by str4t0m** [comment]
Updated the PR today to also move the clock divider from Kconfig to devicetree.

**Comment by erwango** [comment]
@str4t0m Is there a reason not to apply clock config to other boards configuring fdcan: nucleo_h743zi, nucleo_h753zi and stm32h7b3i ?

**Comment by str4t0m** [comment]
> Is there a reason not to apply clock config to other boards configuring fdcan: nucleo_h743zi, nucleo_h753zi and stm32h7b3i ?

Yes, STM32H7 use the `st,stm32h7-fdcan` driver which didn't have these configuration options.
And I don't have access to any such board to verify the changes.

**Review by erwango — DISMISSED** [review]

File: drivers/can/can_stm32fd.c [hunk 1]
```diff
@@ -74,25 +63,29 @@ static int can_stm32fd_clock_enable(const struct device *dev)
...
+				NULL);
+		if (ret < 0) {
+			LOG_ERR("Could not select can_stm32fd domain clock");
+			return ret;
+		}
+	}
 
 	if (!device_is_ready(clk)) {
 		return -ENODEV;
 	}
```
Comment (henrikbrixandersen) [inline drivers/can/can_stm32fd.c:78]: This needs to be moved to before configuring the domain clock, right?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: drivers/can/can_stm32fd.c (hunk 1 above)
Comment (str4t0m) [reply drivers/can/can_stm32fd.c]: Thanks!
Yes it should. While in the current clock control driver implementation it wouldn't really matter it definitely should be checked before being used.

**Review by henrikbrixandersen — APPROVED** [review]

**Review by erwango — APPROVED** [review]

**Pull request merged**